RFC.md

```diff
@@ -0,0 +1,413 @@
+- Implement two initial layers `TransformerSelfAttentionLayer` and `TransformerCrossAttentionLayer`. `TransformerSelfAttentionLayer` would replace `TransformerDecoderLayer`. These would just be two possible layers, but the design would support more variations.
+- Generalize `CausalSelfAttention` work in all three scenarios above (red, green, blue) and rename to `GroupedQueryAttention`. This includes changing the default mask behavior to not always be causal.
+- **Extra**:
+	- Add TanhGate module to support gated attention [ref](https://paperswithcode.com/method/attention-gate)
+	- Modify `TransformerDecoder` to output it's hidden layers and make `model.output` optional to support the changes proposed [#1017](https://github.com/pytorch/torchtune/issues/1017) and make only one BC breaking change.
+	- Implement legacy checkpoint conversion to support old checkpoints
+
+### TransformerDecoder
+TransformerDecoder only highlighting the changes, current module [here](https://github.com/pytorch/torchtune/blob/0057fe7cf83e14f0b62538a8d4d20719e0a88639/torchtune/modules/transformer.py#L99). Discussion in the comments.
+
+```python
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        *,
+        ...
+        # as opposed to layer and num_layers
+        layers: List[nn.Module],
+        ...
+        # output is generalized and hidden_idx is added to address #1017
+        output: Callable
+        hidden_idx: Optional[List[int]] = None
+    ) -> None:
+        super().__init__()
+        # We could keep an optional layer + num_layers args for BC
+        # Otherwise this change requires builders to go from
+        # TransformerDecoder(layer, 32)
+        # To
+        # for _ in range(32):
+	    #    layers.append(layer)
+	    # TransformerDecoder(layers)
+        ...
+
+    def forward(
+        self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> [Tensor, List[Tensor]]:
+        ...
+
+		# Move max seq length check from attention to here
+        if seq_len > self.max_seq_len:
+            raise ValueError("Too long")
+
+        ...
+
+		hidden = []
+        for i, layer in enumerate(self.layers):
+	        if i in self.hidden_idx:
+		        hidden.append(h)
+            h = layer(
+                h,
+                mask=mask,
+                encoder_input=encoder_input,
+                encoder_mask=encoder_mask,
+                input_pos=input_pos,
+            )
+
+        # shape: [b, s, d]
+        h = self.norm(h)
+
+        output = self.output(h).float()
+        # this unifies the output with ViT but breaks recipes
+        # Alternative:
+        # output = output if not hidden else hidden + [output]
+        return output, hidden
+```
+@SalmanMohammadi does this design address all your needs from #1017
```

Comment:
So the only thing we're changing here is the addition of returning hidden states, right?

I think the other half of #1017 was flexibility around how the output from the last hidden layer is used for some downstream task(s), particularly cases which could modify the return signature of `TransformerDecoder.forward`.

I admit since we're omitting joint policy-value head networks in PPO we currently don't have a concrete use case for this, so I'm not really concerned about this atm. If this isn't something that's on your horizon for supporting down the line, then I don't think we actually need #1017.

Reply:
For joint policy heads, couldn't you just define your head as

```python
PolicyValueHead(nn.Module):
	...
def forward(self, h):
	p = self.policy(h)
	v = self.value(h)
	return p, v
```
and then set `output=PolicyValueHead()`? But also if we don't have a concrete use case for it, I'm fine with not worrying about it until we do.

Reply:
Agreed, I don't the added complexity is worth it. Overall this is really great. Really clearly written and presented - thank you : )